### PR TITLE
Creates image proxy view (closes #104).

### LINKED ISSUE #104: Use an image proxy to resize large images and prevent IP address leakage.
Not clear if we can reuse the embedly-proxy for this purpose or not, let's look into it.

### PR DESCRIPTION
**Pull request by chuckharmston**
After discussion with ops, we decided to take a different approach to this bug than #103; instead of standing up our own service, we'll use [Embedly's resize API](http://embed.ly/docs/explore/display/resize), and let nginx's upstream cache manage the caching.

TODO:

- [x] Write tests
- [x] Rewrite images in recommendations to use this.
- [x] Use the `url` querystring param in other images provided by Embedly, e.g. `https://i.embed.ly/1/image?url=https://upload.wikimedia.org/wikipedia/en/thumb/5/53/Captain_America_Civil_War_poster.jpg/220px-Captain_America_Civil_War_poster.jpg&key=%3Ckey%3E`

r? @6a68 @hannosch @mostlygeek

### CODE REVIEW

File: conf/web.sh [hunk 1]
```diff
@@ -1,3 +1,3 @@
 #!/usr/bin/env bash
 
-exec uwsgi --http :${PORT:-8000} --wsgi-file /app/recommendation/wsgi.py --master
+exec uwsgi --wsgi-disable-file-wrapper --http :${PORT:-8000} --wsgi-file /app/recommendation/wsgi.py --master
```
Comment (chuckharmston) [inline conf/web.sh]: This [is necessary](https://www.python.org/dev/peps/pep-0333/#optional-platform-specific-file-handling) to let uwsgi return a streamed `ByteIO` object.

File: recommendation/mozlog/middleware.py [hunk 2]
```diff
@@ -53,7 +56,7 @@ def request_summary(response):
     log['lang'] = request.headers.get('Accept-Language')
     log['method'] = request.method
     log['path'] = request.path
-    log['t'] = (request.finish_time - request.start_time) * 1000 # in ms
+    log['t'] = (request.finish_time - request.start_time) * 1000  # in ms
```
Comment (chuckharmston) [inline recommendation/mozlog/middleware.py]: Drive-by PEP8 fix.

**Comment by mostlygeek** [comment]
looks good so far!

**Comment by 6a68** [comment]
@chuckharmston Would you mind explaining the rationale behind this change? Is it to avoid having every single Universal Search user pinging the Embedly display API?

**Comment by chuckharmston** [comment]
This was a change requested by ops to prevent the risks and overhead
associated with allowing our server to make requests to the public internet.

File: recommendation/views/images.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+
+@images.route('/images')
+def proxy():
+    try:
+        url = make_embedly_url(**request.args)
+    except RuntimeError:
+        abort(400)
+    req = requests.get(url, stream=True)
+    stream = stream_with_context(req.iter_content())
```
Comment (hannosch) [inline recommendation/views/images.py]: This should check for the response status code and only try to get the content if we got a 200 OK response. I couldn't find a clear explanation of all the possible error codes for embedly, but it seems they can at least return various 4xx responses. For example if the API key is invalid or if a url is requested that either doesn't exist or isn't an image.

File: recommendation/views/images.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+
+def make_embedly_url(**kwargs):
+    if 'url' not in kwargs:
+        raise RuntimeError()
+    qs = {}
+    for param in ['url', 'width', 'height']:
+        if param in kwargs:
+            qs[param] = kwargs[param][0]
+    qs['key'] = conf.EMBEDLY_API_KEY
```
Comment (hannosch) [inline recommendation/views/images.py]: I'm not sure, but maybe we should specify ``animate: false`` and ``compresspng: true`` in the outgoing API call.

Turning off animation makes sure that we don't return animated images, but just the first frame of one. I think in the UI context where we want to show these images, animated images would be a distraction.

compresspng lowers the returned file sizes at the expense of a slightly longer response time from embedly. But since both embedly and we are caching the response, this added latency shouldn't be visible to the end users all too often. And if the cache is effective, most of them will benefit from the smaller image sizes.

File: recommendation/views/images.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    return '{}?{}'.format(EMBEDLY_RESIZE, urlencode(qs))
+
+
+@images.route('/images')
+def proxy():
+    try:
+        url = make_embedly_url(**request.args)
+    except RuntimeError:
+        abort(400)
+    req = requests.get(url, stream=True)
```
Comment (hannosch) [inline recommendation/views/images.py]: IIRC requests doesn't set a default timeout for connections. So I'd put something here manually, and also make sure to catch any exception for either timeouts, SSL errors or anything else that can go wrong with outbound calls:

```
try:
    req = requests.get(url, stream=True, timeout=60.0)  # timeout in seconds
except requests.RequestException:
    abort(400)
```

File: recommendation/views/images.py (hunk 3 above)
Comment (hannosch) [inline recommendation/views/images.py]: Regarding my previous comment, maybe something like:
```
if req.status_code != 200:
    abort(400)
stream = stream_with_context(req.iter_content())
```

File: recommendation/mozlog/middleware.py [hunk 6]
```diff
@@ -42,7 +42,10 @@ def request_summary(response):
 
     log = {}
     query = request.args.get('q')
-    data = response.get_data(as_text=True)
+    try:
+        data = response.get_data(as_text=True)
+    except UnicodeDecodeError:
+        data = ''
```
Comment (hannosch) [inline recommendation/mozlog/middleware.py]: The body is only used if we also get a query. What about moving this block into the ``if query:`` block later on, so it doesn't get called for the new image API at all?

File: recommendation/views/images.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+def make_embedly_url(**kwargs):
+    if 'url' not in kwargs:
+        raise RuntimeError()
+    qs = {}
+    for param in ['url', 'width', 'height']:
+        if param in kwargs:
+            qs[param] = kwargs[param][0]
+    qs['animate'] = False
+    qs['compresspng'] = conf.EMBEDLY_API_KEY
```
Comment (hannosch) [inline recommendation/views/images.py]: @chuckharmston copy/paste error, either this should be True or a different config value.

**Comment by 6a68** [comment]
@chuckharmston I'm running short on time before vacation, so I don't think I'll have time to review this patch. @hannosch seems to have bandwidth to provide additional feedback, merge when you feel it's ready

**Comment by chuckharmston** [comment]
Other than the one `except` clause that isn't getting hit by the test that I think should be hitting it, this is ready for review. r? @hannosch @mostlygeek

**Comment by mostlygeek** [comment]
What sort of server side changes are required to deploy this update?
Any new environment variables that need to be set?

**Comment by chuckharmston** [comment]
I'm going to merge this to make sure Jared has it available to him while testing https://github.com/mozilla/universal-search/issues/200. We'll ship it to stage today and do some extra testing before letting it hit prod some time next week.

**Comment by chuckharmston** [comment]
CircleCI seems to be stuck, so by virtue of CI passing Travis, I'm going to disable the merge block for now and merge it anyway. See https://github.com/mozilla/universal-search-recommendation/issues/145

**Comment by hannosch** [comment]
This does look all good to me, so a belated r+
